cpp/src/arrow/io/interfaces.cc

```diff
@@ -133,6 +140,15 @@ Status RandomAccessFile::ReadAt(int64_t position, int64_t nbytes,
 
 Status RandomAccessFile::GetSize(int64_t* size) { return GetSize().Value(size); }
 
+// Default ReadAsync() implementation: simply issue the read on one of the IO threads
+Result<Future<std::shared_ptr<Buffer>>> RandomAccessFile::ReadAsync(int64_t position,
+                                                                    int64_t nbytes) {
+  auto pool = internal::GetIOThreadPool();
```

Comment:
Random thoughts: should the IO thread pool be an argument of `ReadAsync`? Or we could introduce an abstraction like `AsyncContext` (thinking out loud). I guess probably a good thing to mark all async APIs as experimental until a broader pattern can emerge

Reply:
I left more comments about this in my review-level comment

Reply:
I agree that an `AsyncContext` (at least as an optional argument) is useful. Is there a reason this would be more than a wrapper around (or a promotion of) `internal::TaskGroup`?

Reply:
Why not, but some implementations may not use it (for example, for S3 I should probably make use of the AWS SDK's own async APIs).

Reply:
I see. So we'd need something like:
```c++
interface AsyncContext;

/// an AsyncContext which dispatches tasks to a TaskGroup
class TaskGroupAsyncContext : public AsyncContext {
  TaskGroupAsycnContext(TaskGroup);
  MakeThreaded(ThreadPool* = default_thread_pool());
  MakeSerial();
};

/// an AsyncContext which dispatches tasks using AWS SDK's async APIs.
/// Warning: this can only dispatch AwsTasks
class AwsAsyncContext : public AsyncContext {
  Make()
};

interface AwsTask;
```

Reply:
Why would the caller need to pass that information? My point is that even if the caller passes a thread pool (or some other "async" backend), the implementation might choose to ignore it.

Reply:
I don't think this needs to be resolved in this patch, but generally we need to come up with a consistent strategy for externalizing thread scheduling logic in functions that spawn asynchronous tasks